xds/internal/xdsclient/authority.go

```diff
@@ -639,16 +639,23 @@ func (a *authority) watchResource(rType xdsresource.Type, resourceName string, w
 			if a.logger.V(2) {
 				a.logger.Infof("Resource type %q with resource name %q found in cache: %s", rType.TypeName(), resourceName, state.cache.ToJSON())
 			}
+			// state can only be accessed in the context of an
+			// xdsClientSerializer callback. Hence making a copy of the cached
+			// resource here for watchCallbackSerializer.
 			resource := state.cache
 			a.watcherCallbackSerializer.TrySchedule(func(context.Context) { watcher.OnUpdate(resource, func() {}) })
 		}
 		// If last update was NACK'd, notify the new watcher of error
 		// immediately as well.
 		if state.md.Status == xdsresource.ServiceStatusNACKed {
 			if a.logger.V(2) {
-				a.logger.Infof("Resource type %q with resource name %q was NACKed: %s", rType.TypeName(), resourceName, state.cache.ToJSON())
+				a.logger.Infof("Resource type %q with resource name %q was NACKed", rType.TypeName(), resourceName)
 			}
-			a.watcherCallbackSerializer.TrySchedule(func(context.Context) { watcher.OnError(state.md.ErrState.Err, func() {}) })
+			// state can only be accessed in the context of an
+			// xdsClientSerializer callback. Hence making a copy of the error
+			// here for watchCallbackSerializer.
+			err := state.md.ErrState.Err
```

Comment:
For my understanding, why would we need to make a copy of the error for the callback serializer? What is the consequence of just using `state.md.ErrState.Err` directly as previously? Is it for easier testing?

Reply:
`state` must *only* be accessed in the context of an `xdsClientSerializer` callback. The serializer ensures that only one callback is active at any given time, thereby guaranteeing mutual exclusion and therefore we don't need to guard accesses to `state` with a mutex. See: https://github.com/grpc/grpc-go/blob/master/xds/internal/xdsclient/authority.go#L90

If you observer closely, the call to `watcher.OnError` does not happen in the context of an `xdsClientSerializer` callback. It happens in the context of a `watcherCallbackSerializer` callback. So, the race detector was complaining about a data race. Making a copy of the error solves the issue. If you look at line 646, the same is done when `watcher.OnUpdate` is called.